Suggest filename encryption with AES-SIV-512-B64URL (#24)
encryption-alliance/unified-vault-format: file name encryption/AES-SIV-512-B64URL.md

# File Name Encryption using AES-SIV-512 + Base64-URL-Encoding

## Directory Metadata

Every directory requires certain metadata that affects the file name encryption of its direct children:

* the seed used to derive keys
* the directory ID

This data is immutable and therefore linked with a directory eternally, surviving renames/moves. This data is stored in a file called `dir.uvf`, which is stored in two places:
1. Within the parent dir (except for root), where it serves a link to the child dir
2. In the child dir itself (allowing disaster recovery without the parent)

The exact file structure of `dir.uvf` will be discussed in more detail [below](#format-of-diruvf-and-symlinkuvf).

### Directory Seed

> At creation time, a directory's seed is always the `latestSeed` as defined in the [vault metadata file](../vault%20metadata/README.md#sensitive-metadata). In case of the root directory this happens to also be the `initialSeed`.

Reviewer: ```suggestion
At creation time, a directory's seed is always the `latestSeed` as defined in the [vault metadata file](../vault%20metadata/README.md#encrypted-content). In case of the root directory this happens to also be the `initialSeed`.
```

> At creation time, a directory's seed is always the `latestSeed` as defined in the [vault metadata file](../vault%20metadata/README.md#sensitive-metadata). In case of the root directory this happens to also be the `initialSeed`.

Reviewer: @overheadhunter what's the motivation for moving away from using the same `nameKey` eternally to using the `latestSeed` for new directories? Furthermore, afaics, the same `latestSeed` is now used for content encryption and file name encryption in new directories - is there a guarantee that no information can be inferred from the combined use?

Author: this allows key rotation to also be effective for names of nodes added to such directories.

Reviewer: Does key rotation for names improve security in any relevant way? Is it worth the additional lookup/memory overhead in directory traversal of a vault - is this not restrictive?

Author: > Does key rotation for names improve security in any relevant way?

If users pick a product for the ability to encrypt names, they surely don't want names to leak to former holders of old encryption keys. So yes, this is relevant.

> Is it worth the additional lookup/memory overhead in directory traversal of a vault - is this not restrictive?

Imo we get this for free, since we have the different seeds for file contents already anyway. Whether protecting a piece of information is *worth* additional effort depends on whether the leaking name is `document.pdf` or `company acquisition offer by Google.pdf`. Only the user can tell.

> When navigating into a directory, the seed is read from the corresponding `dir.uvf`'s file header.

Reviewer: ```suggestion
When navigating into a directory, the seed is read from the corresponding `dir.uvf`'s file header. More precisely, the header contains the seed ID, and the seed can the be looked up in the `seeds` in [vault metadata file](../vault%20metadata/README.md#encrypted-content).
```

> [!NOTE]

> > A directory's seed is used to derive encryption keys used to encrypt the directory's direct children. Since the seed is immutable, new child nodes added at a later time will use the same encryption keys as old children.

Reviewer: ```suggestion
> A directory's seed is used to derive file name encryption keys used to encrypt the directory's direct children's names. Since the seed is immutable, new child nodes added at a later time will use the same file name encryption keys as old children.
```

```ts
let sivKey = kdf(secret: seed, len: 64, context: "siv")
let hmacKey = kdf(secret: seed, len: 32, context: "hmac")
```

## Mapping Directory IDs to Paths

When traversing directories, the directory ID of a given subdirectory is processed in three steps to determine the storage path inside the vault:

1. Compute the HMAC of the `dirId` using SHA-256 and the `hmacKey`
1. Encoding the hash with Base32 to get a string of printable chars
1. Construct the directory path by resolving substrings of the encoded hash relative to `{vaultRoot}/d/`
    * split of the first two characters of the encoded hash (allowing for a total of 1024 directories within the base directory)
    * use the remaining 30 characters of the encoded hash as the second level directory

```ts
let dirIdHash = hmacSha256(data: dirId, key: hmacKey)
let dirIdString = base32(dirIdHash)
let dirPath = vaultRoot + '/d/' + dirIdString[0..2] + '/' + dirIdString[2..32]
```

> [!NOTE]

> > Due to the nature of hierarchical data structures, travering file trees is an inherently top-down process, allowing the use of one-way hash functions.

Reviewer: ```suggestion
> Due to the nature of hierarchical data structures, traversing file trees is an inherently top-down process, allowing the use of one-way hash functions.
```

# File Name Encryption using AES-SIV-512 + Base64-URL-Encoding

## Directory Metadata

Every directory requires certain metadata that affects the file name encryption of its direct children:

* the seed used to derive keys
* the directory ID

This data is immutable and therefore linked with a directory eternally, surviving renames/moves. This data is stored in a file called `dir.uvf`, which is stored in two places:
1. Within the parent dir (except for root), where it serves a link to the child dir

> 2. In the child dir itself (allowing disaster recovery without the parent)

Reviewer: What does "disaster recovery without the parent" mean? Recovery of only a subtree of the unencrypted directory structure in case some encrypted files corresponding to parent folders are corrupted?

Author: Exactly. Imagine cleartext folder structure `a/b/c/` but sync fails and `a/b/` gets lost. With this information, `?/?/c/` and all its children can still be recovered. The `dirId` required for name decryption would otherwise only be available within the lost parent dir.

Reviewer: Moved to [https://github.com/encryption-alliance/unified-vault-format/pull/24/files#r1728617957](https://github.com/encryption-alliance/unified-vault-format/pull/24/files#r1728617957)

> 2. In the child dir itself (allowing disaster recovery without the parent)

Reviewer: ```suggestion
2. In the child dir itself (allowing disaster recovery without the parent). 

> [!NOTE] Disaster recovery without the parent
> Imagine cleartext folder structure `a/b/c/` but sync fails and `a/b/` gets lost. With this information, `?/?/c/` and 
> all its children can still be recovered. The `dirId` required for name decryption would otherwise only be available 
> within the lost parent dir.
```

When navigating into a directory, the seed is read from the corresponding `dir.uvf`'s file header. More precisely, the header contains the seed ID, and the seed can the be looked up in the `seeds` in [vault metadata file](../vault%20metadata/README.md#encrypted-content).

> [!NOTE]
> A directory's seed is used to derive file name encryption keys used to encrypt the directory's direct children's names. Since the seed is immutable, new child nodes added at a later time will use the same file name encryption keys as old children.
>
> Consequently, key rotation only affects file names of nodes added to newly created directories. Key rotation is ineffective for children added to preexisting directories.

### Directory ID

The directory ID is a unique sequence of 32 random bytes (taking the birthday paradox into account, collision probability is therefore 2^-128).

```ts
let dirId = csprng(len: 32)
```

The only exception to this is the root directory ID, which is deterministically derived from the `initialSeed` using the [KDF](../kdf/README.md):

```ts
let rootDirId = kdf(secret: initialSeed, len: 32, context: "rootDirId")
```

## Deriving Encryption Keys

> All file names are encrypted using AES-SIV, which requires a 512 bit key (which is internally split into two 256 bit AES keys). Furthermore we need a 256bit key for HMAC computations. We use the directory-specific seed from `dir.uvf` and feed it into the [KDF](../kdf/README.md):

Reviewer: ```suggestion
All file names are encrypted using AES-SIV, which requires a 512 bit key (which is internally split into two 256 bit AES keys). Furthermore we need a 256 bit key for HMAC computations. We use the directory-specific seed from `dir.uvf` and feed it into the [KDF](../kdf/README.md):
```

```ts
let sivKey = kdf(secret: seed, len: 64, context: "siv")
let hmacKey = kdf(secret: seed, len: 32, context: "hmac")
```

## Mapping Directory IDs to Paths

When traversing directories, the directory ID of a given subdirectory is processed in three steps to determine the storage path inside the vault:

1. Compute the HMAC of the `dirId` using SHA-256 and the `hmacKey`
1. Truncate the result. Keep the leftmost 160 bits, discard the remaining 96 bits

> 1. Encoding the truncated hash with Base32 to get a string of printable chars

Reviewer: ```suggestion
1. Encode the truncated hash with Base32 to get a string of printable chars
```

1. Construct the directory path by resolving substrings of the encoded hash relative to `{vaultRoot}/d/`
    * split of the first two characters of the encoded hash (allowing for a total of 1024 directories within the base directory)
    * use the remaining 30 characters of the encoded hash as the second level directory

```ts
let dirIdHash = hmacSha256(data: dirId, key: hmacKey)
let truncatedHash = dirIdHash[0..20]
let dirIdString = base32(truncatedHash)
let dirPath = vaultRoot + '/d/' + dirIdString[0..2] + '/' + dirIdString[2..32]
```

> [!NOTE]
> Due to the nature of hierarchical data structures, traversing file trees is an inherently top-down process, allowing the use of one-way hash functions.
>

> > Base32 is used to get a encoding that works with case insensitive file systems and limits the number of nodes within `d/` to `32^2`.

Reviewer: ```suggestion
> Base32 is used to get an encoding that works with case insensitive file systems and limits the number of nodes within `d/` to `32^2`.
```

```txt
.
├─ vault.uvf
└─ d
   ├─ BZ
   │  └─ R4VZSS5PEF7TU3PMFIMON5GJRNBDWA     # Root Directory
   │     ├─ dir.uvf                         # Root Directory's metadata
   │     ├─ 5TyvCyF255sRtfrIv83ucADQ.uvf    # File.txt
   │     ├─ FHTa55bHsUfVDbEb0gTL9hZ8nho.uvf # Subdirectory
   │     │  └─ dir.uvf                      # Subdirectory's metadata
   │     └─ gLeOGMCN358UBf2Qk9cWCQl.uvf     # Symlink
   │        └─ symlink.uvf                  # Symlink's target
   ├─ FC
   │  └─ ZKZRLZUODUUYTYA4457CSBPZXB5A77     # Subdirectory
   │     ├─ dir.uvf                         # Subdirectory's metadata
   |     └─ ...                             # Subdirectory's children
   └─ ...
```

### Traversing the Example Directory Structure

#### List contents of `/`:

1. Use `initialSeed` as a seed for the root directory
1. compute `rootDirId` and corresponding ciphertext dir path -> `d/BZ/R4VZSS5PEF7TU3PMFIMON5GJRNBDWA`
1. list direct children within `d/BZ/R4VZSS5PEF7TU3PMFIMON5GJRNBDWA`
    * `dir.uvf` (file)
    * `5TyvCyF255sRtfrIv83ucADQ.uvf` (file)
    * `FHTa55bHsUfVDbEb0gTL9hZ8nho.uvf` (dir)
    * `gLeOGMCN358UBf2Qk9cWCQl.uvf` (dir)

> 1. For each subdirectory, determine node type

Reviewer: ```suggestion
1. for each subdirectory, determine node type
```

The byte sequence is then encrypted using [AES-SIV](https://tools.ietf.org/html/rfc5297). In order to bind the node to the containing directory, preventing undetected manipulation of the folder structure, the directory ID of the parent folder is used as associated data.

Lastly, the ciphertext is encoded with unpadded base64url.

```txt
ciphertext := aesSiv(secret: cleartextName, ad: parentDirId, key: sivKey)
ciphertextName := base64url(data: ciphertext) + '.uvf'
```

## Ciphertext Directory Structure

Depending on the kind of node, the encrypted name is then either used to create a file or a directory:

* Files are stored as files.
* Non-files are stored as directories. The type of the node then depends on this directory's content.
    * Directories are denoted by a file called `dir.uvf` containing aforementioned directory ID.
    * Symlinks are denoted by a file called `symlink.uvf` containing the encrypted link target.
    * Further types may be appended in future releases.

Thus, a cleartext directory structure like this:

```txt
.
├─ File.txt
├─ SymlinkToFile.txt
├─ Subdirectory
│  └─ ...
└─ ...
```

> The corresponding ciphertext directory structure will then be:

Author: added in 4d1532d3bd76dbf55d2f20a2e3a3106396dc6f7e

```txt
.
├─ d
│  ├─ BZ
│  │  └─ R4VZSS5PEF7TU3PMFIMON5GJRNBDWA
│  │     ├─ dir.uvf  # current dirId
│  │     ├─ 5TyvCyF255sRtfrIv**83ucADQ.uvf  # File.txt
│  │     ├─ FHTa55bH*sUfVDbEb0gTL9hZ8nho.uvf  # Subdirectory
│  │     │  └─ dir.uvf  # subdir dirId
│  │     └─ gLeOGMCN358*UBf2Qk9cWCQl.uvf  # SymlinkToFile.txt
│  │        └─ symlink.uvf  # contains link target
│  └─ FC
│     └─ ZKZRLZUODUUYTYA4457CSBPZXB5A77  # contains contents of Subdirectory
│  │     ├─ dir.uvf  # current dirId
│        └─ ...
└─ vault.uvf

> ```

Author: added in dfc0111502769138e5137cd9ccd29419dad71b1d
